### ISSUE TITLE
sources doesn't include unmanaged source code

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#4650]
## steps

Try 1.3.0-M3 with shapeless. https://github.com/milessabin/shapeless

```
sbt:root> show coreJVM/Compile/sources
```

## problem

It only shows src_managed files, and thus `compile` fails.

## expectation

It lists the source code.

**Comment by eed3si9n** [comment]
Similar issue with specs.

```scala
addSbtPlugin("org.scala-js"       % "sbt-scalajs"              % "0.6.27")
addSbtPlugin("org.portable-scala" % "sbt-crossproject"         % "0.6.0")
addSbtPlugin("org.portable-scala" % "sbt-scalajs-crossproject" % "0.6.0")
...
```

It doesn't find some sources in Compile:

`show commonJVM/Compile/sources` shows only the source from `common/jvm/src` but not `common/shared/src`.

**Comment by eed3si9n** [comment]
```scala
sbt:root> show coreJVM/Compile/unmanagedSources/fileInputs
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala-2.12/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/java/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala-2.12/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala-2.12/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala_2.11+/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala_2.13-/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/java/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala-2.12/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala_2.11+/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala_2.13-/**/*.{java,scala}
[info] * /Users/eed3si9n/work/quicktest/shapeless/core/jvm/*.{java,scala}
sbt:root> show coreJVM/Compile/unmanagedSources/inputFileStamps
[info] *
```

**Comment by eed3si9n** [comment]
```scala
scala> (coreJVM/Compile/unmanagedSources/fileTreeView).eval
res1: sbt.nio.file.FileTreeView[(java.nio.file.Path, sbt.nio.file.FileAttributes)] = sbt.nio.file.FileTreeView$DefaultFileTreeView$@30ba29cf

scala> (coreJVM/Compile/unmanagedSources/fileInputs).eval
res2: Seq[sbt.nio.file.Glob] = List(/Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala-2.12/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/java/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala-2.12/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/src/main/scala/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala-2.12/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala_2.11+/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core/jvm/src/main/scala_2.13-/**/*.{java,scala}, /Users/eed3si9n/work/quicktest/shapeless/core...

scala> res1.list(res2)
res3: Seq[(java.nio.file.Path, sbt.nio.file.FileAttributes)] = Vector()
```

**Comment by eed3si9n** [comment]
I think I have a repro - https://github.com/sbt/io/pull/237

**Comment by eed3si9n** [comment]
Calling distinct on the glob list works around the issue, but I don't know why duplicate has such devastating effect.

**Comment by eed3si9n** [comment]
FTR, I updated https://github.com/sbt/io/pull/237 so the iterator doesn't get stuck when it finds a path that it's seen already.

**Comment by milessabin** [comment]
Anything I can do to help with this?

**Comment by eed3si9n** [comment]
No. I am happy to have found the sbt bug before we released sbt 1.3.0 thanks to Community Build.
